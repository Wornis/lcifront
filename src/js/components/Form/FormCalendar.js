import React from 'react';
import {DatePicker} from 'material-ui-pickers';
import AccessTimeIcon from '@material-ui/icons/AccessTime';
import format from 'date-fns/format';
import PropTypes from 'prop-types';

const FormCalendar = props => {
  const onChange = (newDate) =>
    props.onChangeDate(format(newDate, 'yyyy-MM-dd'));

  return (
    <div className="picker" style={{marginBottom: 15, width: 240}}>
      <AccessTimeIcon style={{marginTop: 20}} />
      <DatePicker
        disabled={props.disabled}
        id='formCalendarDatePicker'
        label="Date"
        value={props.selectedDate}
        style={{marginLeft: 10}}
        onChange={onChange.bind(this)}
        animateYearScrolling
        format="d MMM yyyy"
      />
    </div>
  );
};

FormCalendar.propTypes = {
  selectedDate: PropTypes.string.isRequired,
  onChangeDate: PropTypes.func.isRequired
};
export default FormCalendar;

