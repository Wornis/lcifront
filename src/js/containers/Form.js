import React from 'react';
import PropTypes from 'prop-types';
import Avatar from '@material-ui/core/Avatar';
import Button from '@material-ui/core/Button';
import CreateIcon from '@material-ui/icons/Create';
import CreditCardIcon from '@material-ui/icons/CreditCard';
import EuroSymbolIcon from '@material-ui/icons/EuroSymbol';
import MonetizationOnIcon from '@material-ui/icons/MonetizationOn';
import MoneyIcon from '@material-ui/icons/Money';
import MyLocationIcon from '@material-ui/icons/MyLocation';
import Paper from '@material-ui/core/Paper';
import Typography from '@material-ui/core/Typography';
import withStyles from '@material-ui/core/styles/withStyles';
import TextField from '@material-ui/core/es/TextField/TextField';
import FormControl from '@material-ui/core/es/FormControl/FormControl';
import InputLabel from '@material-ui/core/es/InputLabel/InputLabel';
import Select from '@material-ui/core/es/Select/Select';
import MenuItem from '@material-ui/core/es/MenuItem/MenuItem';
import FormCalendar from '../components/Form/FormCalendar';
import format from 'date-fns/format';
import isValid from 'date-fns/isValid';
import {CircularProgress} from '@material-ui/core/es/index';

const styles = theme => ({
    main: {
        width: 'auto',
        display: 'block', // Fix IE 11 issue.
        marginLeft: theme.spacing.unit * 3,
        marginRight: theme.spacing.unit * 3,
        [theme.breakpoints.up(400 + theme.spacing.unit * 3 * 2)]: {
            width: 400,
            marginLeft: 'auto',
            marginRight: 'auto',
        },
    },
    paper: {
        marginTop: theme.spacing.unit * 8,
        display: 'flex',
        flexDirection: 'column',
        alignItems: 'center',
        padding: `${theme.spacing.unit * 2}px ${theme.spacing.unit * 3}px ${theme.spacing.unit * 3}px`,
    },
    avatar: {
        margin: theme.spacing.unit,
        backgroundColor: theme.palette.secondary.main,
    },
    form: {
        width: '100%', // Fix IE 11 issue.
        marginTop: theme.spacing.unit,
        display: 'flex',
        flexDirection: 'column',
        alignItems: 'center'
    },
    submit: {
        marginTop: theme.spacing.unit * 3,
        width: 150
    },
    textField: {
        marginLeft: theme.spacing.unit,
        marginRight: theme.spacing.unit,
        width: 200
    },
    formControl: {
        margin: theme.spacing.unit,
        width: 200,
    },
    inputColor: {
        color: 'red'
    }
});

const style = {
    icon: {
        marginTop: 20
    },
    divTextField: {
        marginBottom: 30
    }
};

class Form extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            espValue: '',
            trValue: '',
            cbValue: '',
            totalValue: '',
            dateValue: format(new Date(), 'yyyy-MM-dd'),
            place: 'none',
            submitted: false,
            errors: {
                espValue: false,
                trValue: false,
                cbValue: false
            }
        };
    }

    checkIfErrOnMoneyFields = () => {
        const keys = ['espValue', 'trValue', 'cbValue'];
        let emptyFields = false;
        keys.forEach(key => {
            if (!this.state[key]) { // If the value is empty
                emptyFields = true;
                this.setState((prevState) => ({errors: {...prevState.errors, [key]: true}}));
            }
        });
        const errored = Object.values(this.state.errors).some(boolError => boolError); // Check if no errors
        return emptyFields || errored;
    };

    handleSubmit = () => {
        const boolErrMoneys = this.checkIfErrOnMoneyFields();
        if (!boolErrMoneys && this.state.place && isValid(new Date(this.state.dateValue))) {
            return this.setState({submitted: true});
        }
    };

    checkIfInCorrectValue = (value) => {
        if (value.includes('.')) {
            const afterDotValue = value.substring(value.indexOf('.') + 1);
            return (isNaN(value) || afterDotValue.length > 2);
        }
        return isNaN(value);
    };

    getNewTotal = (selectedKey, value) => {
        value = (!value) ? 0 : value;
        const espValue = (selectedKey === 'espValue') ? value : this.state.espValue || 0;
        const trValue = (selectedKey === 'trValue') ? value : (this.state.trValue || 0);
        const cbValue = (selectedKey === 'cbValue') ? value : this.state.cbValue || 0;
        const newTotal = (parseFloat(espValue) + parseFloat(trValue) + parseFloat(cbValue)).toFixed(2);
        return (newTotal > 0) ? newTotal : '';
    };

    onChangeValue = (e) => {
        const selectedKey = e.target.id;
        const value = e.target.value;
        let errored = false;

        if (this.checkIfInCorrectValue(value)) {
            errored = true;
            this.setState((prevState) => ({errors: {...prevState.errors, [selectedKey]: true}}));
        } else if (this.state.errors[selectedKey]) {
            this.setState((prevState) => ({errors: {...prevState.errors, [selectedKey]: false}}));
            if (!value) // If value field is empty
                return this.setState({[selectedKey]: value, totalValue: ''});
        }
        this.setState({[selectedKey]: value});
        if (errored) {
            return this.setState({totalValue: ''});
        }
        return this.setState({totalValue: this.getNewTotal(selectedKey, value)});
    };

    onChangeDate = (dateValue) =>
        this.setState({dateValue});

    getHelperText = (errorValue) =>
        errorValue ? 'La valeur saisie est incorrecte' : '';

    render() {
        const {classes} = this.props;
        return (
            <main className={classes.main}>
                <Paper className={classes.paper}>
                    <Avatar className={classes.avatar}>
                        <CreateIcon/>
                    </Avatar>
                    <Typography component="h1" variant="h5">
                        Formulaire
                    </Typography>
                    <form className={classes.form} autoComplete='off'>
                        <FormCalendar
                            selectedDate={this.state.dateValue}
                            onChangeDate={this.onChangeDate.bind(this)}
                        />
                        <div style={{marginBottom: 20}}>
                            <MyLocationIcon style={{marginTop: 30}}/>
                            <FormControl className={classes.formControl}>
                                <InputLabel htmlFor="select-place">Emplacement</InputLabel>
                                <Select
                                    value={this.state.place}
                                    inputProps={{name: 'place', id: 'select-place'}}
                                    onChange={(e) => this.setState({place: e.target.value})}
                                >
                                    <MenuItem value='none'>None</MenuItem>
                                    <MenuItem value={10}>Ten</MenuItem>
                                    <MenuItem value={20}>Twenty</MenuItem>
                                    <MenuItem value={30}>Thirty</MenuItem>
                                </Select>
                            </FormControl>
                        </div>
                        <div style={style.divTextField}>
                            <EuroSymbolIcon style={style.icon}/>
                            <TextField
                                id="espValue"
                                label="EspÃ¨ces"
                                className={classes.textField}
                                value={this.state.espValue}
                                onChange={this.onChangeValue.bind(this)}
                                error={this.state.errors.espValue}
                                helperText={this.getHelperText(this.state.errors.espValue)}
                            />
                        </div>
                        <div style={style.divTextField}>
                            <MoneyIcon style={style.icon}/>
                            <TextField
                                id="trValue"
                                label="Tickets Restaurants"
                                className={classes.textField}
                                value={this.state.trValue}
                                onChange={this.onChangeValue.bind(this)}
                                error={this.state.errors.trValue}
                                helperText={this.getHelperText(this.state.errors.trValue)}
                            />
                        </div>
                        <div style={style.divTextField}>
                            <CreditCardIcon style={style.icon}/>
                            <TextField
                                id="cbValue"
                                label="Cartes Bancaires"
                                className={classes.textField}
                                value={this.state.cbValue}
                                onChange={this.onChangeValue.bind(this)}
                                error={this.state.errors.cbValue}
                                helperText={this.getHelperText(this.state.errors.cbValue)}
                            />
                        </div>
                        <div style={style.divTextField}>
                            <MonetizationOnIcon style={style.icon}/>
                            <TextField
                                InputProps={{classes: {input: classes.inputColor}}}
                                disabled
                                id="totalValue"
                                label="Total"
                                className={classes.textField}
                                value={this.state.totalValue}
                            />
                        </div>
                        {
                            this.state.submitted ?
                                <CircularProgress className={classes.progress} /> :
                                <Button
                                    type="button"
                                    variant="contained"
                                    color="primary"
                                    className={classes.submit}
                                    onClick={this.handleSubmit.bind(this)}
                                >
                                    Valider
                                </Button>
                        }
                    </form>
                </Paper>
            </main>
        );
    }
}

Form.propTypes = {
    classes: PropTypes.object.isRequired,
};

export default withStyles(styles)(Form);
