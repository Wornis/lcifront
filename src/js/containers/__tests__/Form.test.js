import React from 'react';
import {createShallow} from '@material-ui/core/test-utils';
import Form from '../Form';
import format from "date-fns/format";

/* The dive()method returns the rendered non-DOM child of the current wrapper.
That becomes useful if your component wraps another component in something like a div element,
and what you're interested in testing is that inner component. */
const getFormWrapper = () => createShallow()(<Form/>).dive();

const initialState = {
    espValue: '',
    trValue: '',
    cbValue: '',
    totalValue: '',
    dateValue: format(new Date(), 'yyyy-MM-dd'),
    place: 'none',
    submitted: false,
    errors: {
        espValue: false,
        trValue: false,
        cbValue: false
    }
};

describe('Test Form component', () => {
    it('Renders correctly', () => {
        const wrapper = getFormWrapper();
        //wrapper.setState({dateValue: '2019-04-13'}); // Same selectedDate than snapshot
        expect(wrapper).toMatchSnapshot();
    });

    it('Mount with correct initial state', () => {
        const wrapper = getFormWrapper();
        expect(wrapper.state()).toEqual(initialState);
    });

    it('Successfully handle onChangeValue with correct value', () => {
        const wrapper = getFormWrapper();
        const id = 'espValue';
        const value = '10.21';
        wrapper.find(`#${id}`).simulate('change', {target: {id, value}});
        expect(wrapper.state().errors[id]).toBeFalsy();
        expect(wrapper.state()[id]).toEqual(value);
    });

    it('Successfully handle onChangeValue with incorrect value', () => {
        const wrapper = getFormWrapper();
        const id = 'trValue';
        const value = '10.215'; // Only 2 numbers accepted after '.'
        wrapper.find(`#${id}`).simulate('change', {target: {id, value}});
        expect(wrapper.state().errors[id]).toBeTruthy();
        expect(wrapper.state()[id]).toEqual(value);
    });

    it('Successfully handle onChangeValue with correct value when before value was errored', () => {
        const wrapper = getFormWrapper();
        const id = 'espValue';
        wrapper.find(`#${id}`).simulate('change', {target: {id, value: '10.251'}}); //Incorrect value
        expect(wrapper.state().errors[id]).toBeTruthy();
        wrapper.find(`#${id}`).simulate('change', {target: {id, value: '10.25'}});
        expect(wrapper.state().errors[id]).toBeFalsy();
    });

    it('Successfully handle submit with correct values', () => {
        const wrapper = getFormWrapper();
        const value = '10.25';
        wrapper.find('#espValue').simulate('change', {target: {id: 'espValue', value}});
        wrapper.find('#trValue').simulate('change', {target: {id: 'trValue', value}});
        wrapper.find('#cbValue').simulate('change', {target: {id: 'cbValue', value}});
        wrapper.find('#submit_form').simulate('click');
        const exceptedState = {
            ...initialState,
            espValue: '10.25',
            trValue: '10.25',
            cbValue: '10.25',
            totalValue: '30.75',
            submitted: true
        };
        expect(wrapper.state()).toEqual(exceptedState);
    });

    it('Successfully handle submit with an incorrect value', () => {
        const wrapper = getFormWrapper();
        wrapper.find('#espValue').simulate('change', {target: {id: 'espValue', value: '10.25'}});
        wrapper.find('#trValue').simulate('change', {target: {id: 'trValue', value: '10.25'}});
        wrapper.find('#cbValue').simulate('change', {target: {id: 'cbValue', value: '10.251'}}); //Incorrect value
        wrapper.find('#submit_form').simulate('click');
        const exceptedState = {
            ...initialState,
            espValue: '10.25',
            trValue: '10.25',
            cbValue: '10.251',
            totalValue: '',
            errors: {
                espValue: false,
                trValue: false,
                cbValue: true
            }

        };
        expect(wrapper.state()).toEqual(exceptedState);
    });

    it('Successfully handle submit with an empty value', () => {
        const wrapper = getFormWrapper();
        wrapper.find('#espValue').simulate('change', {target: {id: 'espValue', value: '10.25'}});
        wrapper.find('#trValue').simulate('change', {target: {id: 'trValue', value: '10.25'}});
        wrapper.find('#cbValue').simulate('change', {target: {id: 'cbValue', value: ''}}); //Empty value
        wrapper.find('#submit_form').simulate('click');
        const exceptedState = {
            ...initialState,
            espValue: '10.25',
            trValue: '10.25',
            cbValue: '',
            totalValue: '20.50',
            errors: {
                espValue: false,
                trValue: false,
                cbValue: true
            }

        };
        expect(wrapper.state()).toEqual(exceptedState);
    });

    it('Successfully handle onDateChange', () =>  {
        const wrapper = getFormWrapper();
        const formattedDate = format(new Date(2018, 3, 28), 'yyyy-MM-dd');
        wrapper.instance().onChangeDate(formattedDate);
        expect(wrapper.state().dateValue).toEqual(formattedDate)
    });

    it('Successfully handle onChange of Select component of place', () =>  {
        const wrapper = getFormWrapper();
        wrapper.find('#select_form').simulate('change', {target: {value: 'Bièvres'}});
        expect(wrapper.state().place).toEqual('Bièvres')
    });

});


