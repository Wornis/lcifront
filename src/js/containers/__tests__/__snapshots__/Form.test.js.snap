// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Form component Renders correctly 1`] = `
<main
  className="Form-main-1"
>
  <WithStyles(Paper)
    className="Form-paper-2"
  >
    <WithStyles(Avatar)
      className="Form-avatar-3"
    >
      <pure(CreateIcon) />
    </WithStyles(Avatar)>
    <WithStyles(Typography)
      component="h1"
      variant="h5"
    >
      Formulaire
    </WithStyles(Typography)>
    <form
      autoComplete="off"
      className="Form-form-4"
    >
      <FormCalendar
        onChangeDate={[Function]}
        selectedDate="2019-04-10"
      />
      <div
        style={
          Object {
            "marginBottom": 20,
          }
        }
      >
        <pure(MyLocationIcon)
          style={
            Object {
              "marginTop": 30,
            }
          }
        />
        <WithStyles(FormControl)
          className="Form-formControl-7"
        >
          <WithStyles(WithFormControlContext(InputLabel))
            htmlFor="select-place"
          >
            Emplacement
          </WithStyles(WithFormControlContext(InputLabel))>
          <WithStyles(WithFormControlContext(Select))
            id="select_form"
            inputProps={
              Object {
                "id": "select-place",
                "name": "place",
              }
            }
            onChange={[Function]}
            value="none"
          >
            <WithStyles(MenuItem)
              value="none"
            >
              None
            </WithStyles(MenuItem)>
            <WithStyles(MenuItem)
              value={10}
            >
              Ten
            </WithStyles(MenuItem)>
            <WithStyles(MenuItem)
              value={20}
            >
              Twenty
            </WithStyles(MenuItem)>
            <WithStyles(MenuItem)
              value={30}
            >
              Thirty
            </WithStyles(MenuItem)>
          </WithStyles(WithFormControlContext(Select))>
        </WithStyles(FormControl)>
      </div>
      <div
        style={
          Object {
            "marginBottom": 30,
          }
        }
      >
        <pure(EuroSymbolIcon)
          style={
            Object {
              "marginTop": 20,
            }
          }
        />
        <TextField
          className="Form-textField-6"
          error={false}
          helperText=""
          id="espValue"
          label="EspÃ¨ces"
          onChange={[Function]}
          required={false}
          select={false}
          value=""
          variant="standard"
        />
      </div>
      <div
        style={
          Object {
            "marginBottom": 30,
          }
        }
      >
        <pure(MoneyIcon)
          style={
            Object {
              "marginTop": 20,
            }
          }
        />
        <TextField
          className="Form-textField-6"
          error={false}
          helperText=""
          id="trValue"
          label="Tickets Restaurants"
          onChange={[Function]}
          required={false}
          select={false}
          value=""
          variant="standard"
        />
      </div>
      <div
        style={
          Object {
            "marginBottom": 30,
          }
        }
      >
        <pure(CreditCardIcon)
          style={
            Object {
              "marginTop": 20,
            }
          }
        />
        <TextField
          className="Form-textField-6"
          error={false}
          helperText=""
          id="cbValue"
          label="Cartes Bancaires"
          onChange={[Function]}
          required={false}
          select={false}
          value=""
          variant="standard"
        />
      </div>
      <div
        style={
          Object {
            "marginBottom": 30,
          }
        }
      >
        <pure(MonetizationOnIcon)
          style={
            Object {
              "marginTop": 20,
            }
          }
        />
        <TextField
          InputProps={
            Object {
              "classes": Object {
                "input": "Form-inputColor-8",
              },
            }
          }
          className="Form-textField-6"
          disabled={true}
          id="totalValue"
          label="Total"
          required={false}
          select={false}
          value=""
          variant="standard"
        />
      </div>
      <WithStyles(Button)
        className="Form-submit-5"
        color="primary"
        id="submit_form"
        onClick={[Function]}
        type="button"
        variant="contained"
      >
        Valider
      </WithStyles(Button)>
    </form>
  </WithStyles(Paper)>
</main>
`;
